File: Design-Patterns/Builder-Design-Pattern/java/src/builderdesignpattern/DentistEngineer.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package builderdesignpattern;

/**
 *
 * @author yousef
 */
public class DentistEngineer {
    private DentistBuilder dentistBuilder;
    
    public DentistEngineer(DentistBuilder dentistBuilder) {
        this.dentistBuilder = dentistBuilder;
    }
    
    public Dentist getDentist() {
        return this.dentistBuilder.getDentist();
    }
    
    public void constructDentist() {
        this.dentistBuilder.buildPersonalInformation();
        this.dentistBuilder.buildEducation();
        this.dentistBuilder.buildAddress();
        if(this.dentistBuilder instanceof SpecialistDentist)
            this.dentistBuilder.buildSpecializationArea();
    }
}

File: Design-Patterns/Builder-Design-Pattern/php/DentistBuilder/DentistEngineer.php
<?php
 

class DentistEngineer {
    private $dentistBuilder;
    
    public function __construct($dentistBuilder) {
        $this->dentistBuilder = $dentistBuilder;
    }
    
    public function getDentist() {
        return $this->dentistBuilder.getDentist();
    }
    
    public function constructDentist() {
        $this->dentistBuilder.buildPersonalInformation();
        $this->dentistBuilder.buildEducation();
        $this->dentistBuilder.buildAddress();
        if($this->dentistBuilder instanceof SpecialistDentist)
            $this->dentistBuilder.buildSpecializationArea();
    }
}
